### PR TITLE
JAMES-2162 JMAP Mailboxes "sharedWith" property.

### PR DESCRIPTION
**Pull request by chibenwa**
Please ignore MAILBOX-307 commits from https://github.com/linagora/james-project/pull/989

Beware! Commit order is messed up on Github due to rebase order...

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/probe/MailboxProbe.java [hunk 1]
```diff
@@ -54,4 +54,6 @@ ComposedMessageId appendMessage(String username, MailboxPath mailboxPath, InputS
     void reIndexAll() throws Exception;
 
     Collection<String> listSubscriptions(String user) throws Exception;
+
+    void addRights(String namespace, String user, String mailboxName, String targetUser, String rights) throws MailboxException;
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/probe/MailboxProbe.java]: I want to follow on https://github.com/linagora/james-project/pull/985#discussion_r141236276 : 
We mix JMX stuff and interfaces we use pervasively, so we end up with bad API.
IMO, we need a JMX layer that take strings and build objects to call the API (like we do for other protocols)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        Lookup('l'),
+        Read('r'),
+        Seen('s'),
+        T_Delete('t'),
+        Write('w');
+
+        private final char imapRight;
+
+        Right(char imapRight) {
+            this.imapRight = imapRight;
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: why Jmap has an imap field ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    private static Builder toRightsBuilder(Map.Entry<EntryKey, MailboxACL.Rfc4314Rights> entry) {
+        return builder().delegateTo(
+            new Username(entry.getKey().getName()),
+            fromACL(entry.getValue()));
+    }
+
+    private static List<Right> fromACL(MailboxACL.Rfc4314Rights rights) {
+        return rights.list()
+            .stream()
+            .map(MailboxACL.Right::asCharacter)
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: we should not use 'char' value to map from Rfc4314Rights to Jmap Rights.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java [hunk 4]
```diff
@@ -1154,6 +1158,273 @@ public void setMailboxesShouldReturnUpdatedWhenNameUpdateAskedOnExistingMailbox(
...
+    }
+
+    @Test
+    public void updateShouldModifyStoredDataWhenUpdatingACL() {
+        String myBox = "myBox";
+        String user = "user";
+        jmapServer.getProbe(MailboxProbeImpl.class).createMailbox(MailboxConstants.USER_NAMESPACE, username, myBox);
+        Mailbox mailbox = jmapServer.getProbe(MailboxProbeImpl.class).getMailbox(MailboxConstants.USER_NAMESPACE, username, "myBox");
+        String mailboxId = mailbox.getMailboxId().serialize();
+        String requestBody1 =
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: why did you extract to variables ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java [hunk 5]
```diff
@@ -210,6 +209,9 @@ private boolean mustChangeParent(Optional<MailboxId> currentParentId, MailboxId
     private void updateMailbox(Mailbox mailbox, MailboxUpdateRequest updateRequest, MailboxSession mailboxSession) throws MailboxException {
         MailboxPath originMailboxPath = mailboxManager.getMailbox(mailbox.getId(), mailboxSession).getMailboxPath();
         MailboxPath destinationMailboxPath = computeNewMailboxPath(mailbox, originMailboxPath, updateRequest, mailboxSession);
+        if (updateRequest.getSharedWith().isPresent()) {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: please use ifPresent instead

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java [hunk 6]
```diff
@@ -190,6 +211,28 @@ public Rights(Multimap<Username, Right> rights) {
         return rights.asMap();
     }
 
+    public MailboxACL toMailboxAcl() {
+        BinaryOperator<MailboxACL> union = Throwing.binaryOperator(MailboxACL::union);
+
+        return rights.asMap()
+            .entrySet()
+            .stream()
+            .map(entry -> new MailboxACL(
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: we lack a more fluent constructor for MailboxACL that would not require building a map

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java [hunk 7]
```diff
@@ -190,6 +211,28 @@ public Rights(Multimap<Username, Right> rights) {
...
+                ImmutableMap.of(
+                    EntryKey.createUser(entry.getKey().value),
+                    toMailboxAclRights(entry.getValue()))))
+            .reduce(new MailboxACL(), union);
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: MailboxACL.empty() ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/probe/MailboxProbe.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/probe/MailboxProbe.java]: That's another biiiig refactoring, isn't it?

What about splitting to a dedicated ACL probe? Then we can ommit exposing it on JMX, and use types ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java (hunk 6 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: We have...

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java (hunk 4 above)
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: Being consistent with the rest of the class.

I inlined them.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java (hunk 5 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java]: It makes MailboxException complicated and is not consistant with the rest of the method. Maybe enough refactoring for one ticket?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        return true;
+    }
+
+    private static boolean isSupported(EntryKey key) {
+        if (key.isNegative()) {
+            LOGGER.warn("Negative keys are not supported");
+            return false;
+        }
+        if (key.getNameType() != MailboxACL.NameType.user) {
+            LOGGER.warn(key.getNameType() + " is not sopported. Onlu 'user' is.");
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: s/Onlu/Only

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java [hunk 9]
```diff
@@ -74,6 +89,7 @@ public static Right forChar(char c) {
     public static class Username {
         private final String value;
 
+     //   @JsonCreator
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: bad comment?

File: server/container/guice/mailbox/src/main/java/org/apache/james/modules/MailboxProbeImpl.java [hunk 10]
```diff
@@ -192,4 +195,13 @@ public void reIndexAll() throws Exception {
         return subscriptionManager.subscriptions(mailboxSession);
     }
 
+    @Override
+    public void addRights(String namespace, String user, String mailboxName, String targetUser, String rights) throws MailboxException {
```
Comment (rouazana) [inline server/container/guice/mailbox/src/main/java/org/apache/james/modules/MailboxProbeImpl.java]: replaceRights?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMailboxesMethodTest.java [hunk 11]
```diff
@@ -69,6 +69,8 @@
     private static final String NAME = "[0][0]";
     private static final String ARGUMENTS = "[0][1]";
     private static final String FIRST_MAILBOX = ARGUMENTS + ".list[0]";
+    public static final String MAILBOX_NAME = FIRST_MAILBOX + ".name";
+    public static final String SHARED_WITH = FIRST_MAILBOX + ".sharedWith";
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMailboxesMethodTest.java]: I don't think these constants are useful. I think matthieu was more talking about replacing these ugly ImmutableList.of("e", "r", "s", "t", "w")

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java [hunk 12]
```diff
@@ -1154,6 +1158,273 @@ public void setMailboxesShouldReturnUpdatedWhenNameUpdateAskedOnExistingMailbox(
...
+            .post("/jmap")
+        .then()
+            .statusCode(200)
+            .body(NAME, equalTo("mailboxes"))
+            .body(FIRST_MAILBOX + ".name", equalTo(myBox))
+            .body(FIRST_MAILBOX + ".sharedWith", is(ImmutableMap.of()));
+    }
+
+    @Test
+    public void updateShouldModifyStoredDataWhenSwithingACLUser() {
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: Swithing?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java [hunk 13]
```diff
@@ -1154,6 +1158,273 @@ public void setMailboxesShouldReturnUpdatedWhenNameUpdateAskedOnExistingMailbox(
...
+                "  ]" +
+                "]";
+
+        with()
+            .header("Authorization", this.accessToken.serialize())
+            .body(requestBody1)
+            .post("/jmap");
+
+        with()
+            .header("Authorization", this.accessToken.serialize())
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: you can remove these this.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/json/ObjectMapperFactory.java [hunk 14]
```diff
@@ -102,6 +106,24 @@ public void serialize(MailboxId value, JsonGenerator gen, SerializerProvider ser
...
+            return new Rights.Username(key);
+        }
+    }
+
+    public static class RightDeserializer extends JsonDeserializer<Rights.Right> {
+
+        @Override
+        public Rights.Right deserialize(JsonParser p, DeserializationContext ctxt) throws IOException, JsonProcessingException {
+            String nodeValue = p.getValueAsString();
+            Preconditions.checkArgument(nodeValue.length() == 1);
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/json/ObjectMapperFactory.java]: what happens else? is there an integration test for that?

Comment (rouazana) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/json/ObjectMapperFactory.java]: idem when right is unknown

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java (hunk 9 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: Removed in fixup

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java (hunk 12 above)
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: Yep. Can not come up with a better naming for this.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java (hunk 13 above)
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: Funny, I quickly investigated the same things.

No I can not! It's not part of the request definition (because you use request specification to query the authentication endpoint).

That would need a reallocation, and moreover is done nowhere else in JMAP integration tests. I chose concistency. Also thought "too much refactoring" in a single week.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMailboxesMethodTest.java (hunk 11 above)
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMailboxesMethodTest.java]: Is that your comment @mbaechler  ?

Because as such I struggle to find  something more readable than the **ugly** ImmutableList.of(...). At least it clearly displays the rights, and do not requires inspections to know what it means.

If you prefer a constant named RIGHTS_ERSTW with the list extracted, well, I don't.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/json/ObjectMapperFactory.java (hunk 14 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/json/ObjectMapperFactory.java]: I added unit tests on deserialization layer.

Also added integration tests. It gets rejected with a invalidArguments error type in the JMAP layer.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMailboxesMethodTest.java (hunk 11 above)
Comment (mbaechler) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMailboxesMethodTest.java]: Ok, let me write some code snippet, I suggest : 
```
ImmutableList.of(Expunge, Read, blabla)
```

Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMailboxesMethodTest.java]: Ok with Expunge = "e" as a constant, right?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java (hunk 12 above)
Comment (rouazana) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: I think at least you miss a 'c'

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java (hunk 13 above)
Comment (rouazana) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: sorry I don't understand why this.accessToken is different from accessToken here?

Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: Ah yes of course I can remove this, not the access token.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMailboxesMethodTest.java (hunk 11 above)
Comment (rouazana) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMailboxesMethodTest.java]: just need to revert this commit, and I think we are good

Comment (rouazana) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMailboxesMethodTest.java]: done + squashing

**Comment by rouazana** [comment]
test this please

**Comment by rouazana** [comment]
test this please

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RightsTest.java [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+    @Test
+    public void toMailboxAclShouldReturnEmptyAclWhenEmpty() {
+        Rights rights = Rights.EMPTY;
+
+        assertThat(rights.toMailboxAcl())
+            .isEqualTo(new MailboxACL());
+    }
+
+    @Test
+    public void toMailboxActShouldReturnActConversion() throws Exception {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RightsTest.java]: ACL

**Comment by chibenwa** [comment]
🍏  here. No more test neded.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+
+public class Rights {
+    public enum Right {
+        Administer(MailboxACL.Right.Administer),
+        Expunge(MailboxACL.Right.PerformExpunge),
+        Insert(MailboxACL.Right.Insert),
+        Lookup(MailboxACL.Right.Lookup),
+        Read(MailboxACL.Right.Read),
+        Seen(MailboxACL.Right.WriteSeenFlag),
+        T_Delete(MailboxACL.Right.DeleteMessages),
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: why T_ ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+        Right(MailboxACL.Right right) {
+            this.right = right;
+        }
+
+        @JsonValue
+        public char asCharacter() {
+            return right.asCharacter();
+        }
+
+        public MailboxACL.Right getRight() {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: ambiguous name. Maybe toMailboxRight ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+            return right;
+        }
+
+        public static Optional<Right> forRight(MailboxACL.Right right) {
+            return OptionalUtils.ifEmpty(
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: there's a semantic problem with this ifEmpty : you don't expect a "ifSomething execute my lambda" to actually return the optional

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+            .collect(Guavate.toImmutableList());
+    }
+
+    private static boolean isSupported(EntryKey key) {
+        if (key.isNegative()) {
+            LOGGER.warn("Negative keys are not supported");
+            return false;
+        }
+        if (key.getNameType() != MailboxACL.NameType.user) {
+            LOGGER.warn(key.getNameType() + " is not supported. Only 'user' is.");
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: i'm sure I don't want to have a lot of warning in production because of that

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+
+    public static final Rights EMPTY = new Rights(ArrayListMultimap.create());
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(Rights.class);
+
+    private final Multimap<Username, Right> rights;
+
+    @JsonCreator
+    public Rights(Map<Username, List<Right>> rights) {
+        this(rights.entrySet()
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: It looks like you coded a "Map of list to Multimap" method ? Why don't you name it that way ? And why it's not reusable ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java [hunk 21]
```diff
@@ -1154,6 +1164,344 @@ public void setMailboxesShouldReturnUpdatedWhenNameUpdateAskedOnExistingMailbox(
...
+            .post("/jmap")
+        .then()
+            .statusCode(200)
+            .body(NAME, equalTo("error"))
+            .body(ARGUMENTS + ".type", is("invalidArguments"))
+            .body(ARGUMENTS + ".description", containsString("No matching right for 'z'"));
+    }
+
+    @Test
+    public void updateShouldCreateStoredDataWhenSettingNewACL() {
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: shouldApply is simpler, don't you think ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java (hunk 16 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: To indicate it's "t" flag.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java (hunk 18 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: What name would you propose?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java (hunk 19 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: Let's put it in info

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java (hunk 20 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: Guava has some methods for this

Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: Used GUAVA's methods for this

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java (hunk 21 above)
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: I don't understand this one. I can not come with a meaningful name using your suggested improvment.

Wouldn't you mind being more specific?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java (hunk 16 above)
Comment (mbaechler) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: do we care ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java (hunk 18 above)
Comment (mbaechler) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: I didn't said the name is wrong. It should nonetheless return void.
So you would have two separated instructions (3 steps) : 
1. transformation
2. ifEmpty + log
3. return

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java (hunk 21 above)
Comment (mbaechler) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: I don't really like "shouldCreateStoredData" because it's not what the caller really expect.
It's why I proposed shouldApply because I find updateShouldApplyWhenSettingNewACL a bit better.
I know it's not a very good name but at least I find it better than previous one.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java (hunk 20 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: That's buggy. I extracted and tested this.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java (hunk 18 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: This method was first intended to write one liners more easily (originaly used for logging empty optionals with Cassandra related code). Returning the original optional value makes it possible.

I propose **peekIfEmpty** as a name.

Comment (mbaechler) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: it makes no sense to me : what do you peek then ? Maybe the problem is about trying to be too generic. Given we already have peek, orElse, orElseThrow, what problems ifEmpty method solves except logging ?
If we agree it's only useful for logging, we could have a special case for that : logIfEmpty that takes a lambda.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Multimap;
+
+public class Rights {
+    public enum Right {
```
Comment (strokyl) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: What does this class add to MailboxACL.Right ? 
They look so close I think we should have only one class no ?

Comment (mbaechler) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: It's really not a good idea to use objects from a given in another domain just because they are similar (right now).
The first reason is : maybe they are not (ok, in this case, we designed JMAP rights to map IMAP ones).
The second reason is : it doesn't cost much and will guarantee you won't break one domain when changing the other one
It's a common mistake to overvalue re-usability over responsibility separation

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java (hunk 18 above)
Comment (strokyl) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: In one hand, I agree @mbaechler ``ifEmpty`` should be a void but in the other hand it's so conveniant it just return the value.
``peekIfEmpty`` would be definitly worst. 
What about ``identityWithSideEffectIfEmpty`` :trollface:

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+
+        public static Optional<Right> forRight(MailboxACL.Right right) {
+            return OptionalUtils.ifEmpty(
+                Arrays.stream(values())
+                    .filter(jmapRight -> jmapRight.right == right)
+                    .findAny(),
+                () -> LOGGER.warn("Non handled right '" + right + "'"));
+        }
+
+        public static Right forChar(char c) {
```
Comment (strokyl) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: I would have put that in Right class and not in the Rights class

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java (hunk 18 above)
Comment (rouazana) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: or simply executeIfEmpty?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java (hunk 22 above)
Comment (strokyl) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java]: Good explanation thank you

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraBlobsDAO.java [hunk 24]
```diff
@@ -168,7 +168,7 @@ private PreparedStatement prepareInsertPart(Session session) {
 
     private byte[] concatenateDataParts(Stream<BlobPart> blobParts) {
         ImmutableList<byte[]> parts = blobParts
-            .map(blobPart -> OptionalUtils.ifEmpty(
+            .map(blobPart -> OptionalUtils.peekOnEmpty(
```
Comment (strokyl) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraBlobsDAO.java]: Is it just me or this name is really worst than `ifEmpty`

File: server/container/util-java8/src/test/java/org/apache/james/util/GuavaUtilsTest.java [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+            .build())
+            .asMap())
+            .isEqualTo(ImmutableMap.of());
+    }
+
+    @Test
+    public void toMultimapShouldAcceptSingleValuesMaps() {
+        assertThat(GuavaUtils.toMultimap(ImmutableMap
+            .<String, List<String>>builder()
+            .put("k1", ImmutableList.of("v1"))
```
Comment (strokyl) [inline server/container/util-java8/src/test/java/org/apache/james/util/GuavaUtilsTest.java]: Could you add a test with a key associated to an empty list ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraBlobsDAO.java (hunk 24 above)
Comment (mbaechler) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraBlobsDAO.java]: I agree

File: server/container/util-java8/src/test/java/org/apache/james/util/GuavaUtilsTest.java (hunk 25 above)
Comment (mbaechler) [reply server/container/util-java8/src/test/java/org/apache/james/util/GuavaUtilsTest.java]: I'll do

File: server/container/util-java8/src/test/java/org/apache/james/util/GuavaUtilsTest.java [hunk 26]
```diff
@@ -68,4 +68,13 @@ public void toMultimapShouldAcceptMultiplesValuesMaps() {
                 .asMap());
     }
 
+    @Test
+    public void toMultimapShouldHandleCorrectlyEmptyList() {
```
Comment (mbaechler) [inline server/container/util-java8/src/test/java/org/apache/james/util/GuavaUtilsTest.java]: shouldStripEntriesWithEmptyList ?

Comment (strokyl) [reply server/container/util-java8/src/test/java/org/apache/james/util/GuavaUtilsTest.java]: Done
